File: bridge/src/pmd3_bridge/schemas.py
# Copyright 2026 Marcelo Cantos
# SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


# ── Request models ─────────────────────────────────────────────────────────────

class UdidRequest(BaseModel):
    udid: str


class ListAppsRequest(BaseModel):
    udid: str


class LaunchAppRequest(BaseModel):
    udid: str
    bundle_id: str


class KillAppRequest(BaseModel):
    udid: str
    bundle_id: str


class PidForBundleRequest(BaseModel):
    udid: str
    bundle_id: str


class AppStateRequest(BaseModel):
    udid: str
    bundle_id: str


class BatteryRequest(BaseModel):
    udid: str


class ScreenshotRequest(BaseModel):
    udid: str


class CrashReportsListRequest(BaseModel):
    udid: str
    since_iso8601: Optional[str] = None
    process: Optional[str] = None


class SyslogRequest(BaseModel):
    """Streaming syslog watch request.

    pid:           filter to a single process id (-1 = all).
    process_name:  filter by image_name (case-sensitive exact match).
                   Empty = no filter.
    subsystem:     filter by SyslogLabel.subsystem (case-sensitive exact
                   match). Empty = no filter.
    """

    udid: str
    pid: int = -1
    process_name: Optional[str] = None
    subsystem: Optional[str] = None


class CrashReportsPullRequest(BaseModel):
    udid: str
    name: str


class AcquirePowerAssertionRequest(BaseModel):
    udid: str
    type: str
    name: str
    timeout_sec: int
    details: Optional[str] = None


class RefreshPowerAssertionRequest(BaseModel):
    handle_id: str
    timeout_sec: int


class ReleasePowerAssertionRequest(BaseModel):
    handle_id: str


# ── Response models ────────────────────────────────────────────────────────────

class DeviceInfo(BaseModel):
    udid: str
    name: str
    product_type: str
    os_version: str


class ListDevicesResponse(BaseModel):
    devices: list[DeviceInfo]


class AppInfo(BaseModel):
    bundle_id: str
    name: Optional[str] = None
    version: Optional[str] = None


class ListAppsResponse(BaseModel):
    apps: list[AppInfo]


class LaunchAppResponse(BaseModel):
    pid: int


class KillAppResponse(BaseModel):
    pass


class PidForBundleResponse(BaseModel):
    pid: Optional[int] = None


class AppStateResponse(BaseModel):
    """Lifecycle state of one app on the device.

    state is one of:
      - "running":      the app is foregrounded (BKS state_description = Running).
      - "backgrounded": the app exists but is suspended or in background mode
                        (state_description ∈ {Suspended, Background, ...}).
      - "terminated":   the app is not in the BackBoard state list — either
                        never launched in this boot, or iOS has fully reaped it.

    Used by autoawake to detect user-initiated opt-out: a Running →
    backgrounded transition for KeepAwake means the user swiped away
    (or launched another app), and autoawake should stay passive.
    """

    state: str
    description: str = ""  # raw state_description from BKS, for debugging


class BatteryResponse(BaseModel):
    level: Optional[float] = None
    charging: Optional[bool] = None


class ScreenshotResponse(BaseModel):
    png_b64: str


class CrashReportEntry(BaseModel):
    name: str
    process: str
    timestamp: str


class SyslogEntry(BaseModel):
    """One emitted syslog line, structured.

    timestamp is RFC3339 with the device-local timezone preserved (pmd3
    surfaces a tz-aware datetime; we keep that).
    process is the image_name (executable basename) reported by pmd3.
    level is the SyslogLogLevel enum name (NOTICE, INFO, DEBUG,
    USER_ACTION, ERROR, FAULT) — not lower-cased; clients normalise.
    subsystem and category are flattened from SyslogLabel for
    convenience; both empty when label is None.
    """

    pid: int
    timestamp: str
    level: str
    process: str
    subsystem: str = ""
    category: str = ""
    message: str


class CrashReportsListResponse(BaseModel):
    reports: list[CrashReportEntry]


class CrashReportsPullResponse(BaseModel):
    content: str
    mime: str = "application/x-apple-crashreport"


class AcquirePowerAssertionResponse(BaseModel):
    handle_id: str


class RefreshPowerAssertionResponse(BaseModel):
    pass


class ReleasePowerAssertionResponse(BaseModel):
    pass


class DevicePowerStateRequest(BaseModel):
    udid: str


class DevicePowerStateResponse(BaseModel):
    """Power/display state of a device.

    state values:
      "awake"       — display is on and the framebuffer is readable (DVT screenshot succeeded).
      "display_off" — screenshot succeeded but pixels are entirely dark (framebuffer
                      blank, display off or screen saver).
      "asleep"      — screenshot failed with an error that indicates the display/device
                      is off or asleep (e.g. DVT instrument closed by device).
      "unknown"     — could not determine state (tunneld unavailable, developer mode
                      off, or unrecognised error).
    """
    state: str
    detail: Optional[str] = None


class ErrorResponse(BaseModel):
    error: str
    message: str

File: internal/pmd3bridge/schemas.go
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

package pmd3bridge

// DeviceInfo describes a connected iOS device as returned by /v1/list_devices.
type DeviceInfo struct {
	UDID        string `json:"udid"`
	Name        string `json:"name"`
	ProductType string `json:"product_type"`
	OSVersion   string `json:"os_version"`
}

// AppInfo describes an installed application as returned by /v1/list_apps.
type AppInfo struct {
	BundleID string  `json:"bundle_id"`
	Name     *string `json:"name,omitempty"`
	Version  *string `json:"version,omitempty"`
}

// Battery describes battery state as returned by /v1/battery.
type Battery struct {
	Level    *float64 `json:"level,omitempty"`
	Charging *bool    `json:"charging,omitempty"`
}

// CrashReport is one entry from /v1/crash_reports_list.
type CrashReport struct {
	Name      string `json:"name"`
	Process   string `json:"process"`
	Timestamp string `json:"timestamp"`
}

// --- request types (internal; used by Client methods) ---

type listDevicesRequest struct{}

type listAppsRequest struct {
	UDID string `json:"udid"`
}

type launchAppRequest struct {
	UDID     string `json:"udid"`
	BundleID string `json:"bundle_id"`
}

type launchAppResponse struct {
	PID int `json:"pid"`
}

type killAppRequest struct {
	UDID     string `json:"udid"`
	BundleID string `json:"bundle_id"`
}

type pidForBundleRequest struct {
	UDID     string `json:"udid"`
	BundleID string `json:"bundle_id"`
}

type pidForBundleResponse struct {
	PID *int `json:"pid"`
}

type batteryRequest struct {
	UDID string `json:"udid"`
}

type screenshotRequest struct {
	UDID string `json:"udid"`
}

type screenshotResponse struct {
	PNGBase64 string `json:"png_b64"`
}

type crashReportsListRequest struct {
	UDID     string  `json:"udid"`
	SinceISO *string `json:"since_iso8601,omitempty"`
	Process  *string `json:"process,omitempty"`
}

type crashReportsListResponse struct {
	Reports []CrashReport `json:"reports"`
}

type crashReportsPullRequest struct {
	UDID string `json:"udid"`
	Name string `json:"name"`
}

type crashReportsPullResponse struct {
	Content string `json:"content"`
}

type acquirePowerAssertionRequest struct {
	UDID       string  `json:"udid"`
	Type       string  `json:"type"`
	Name       string  `json:"name"`
	TimeoutSec int     `json:"timeout_sec"`
	Details    *string `json:"details,omitempty"`
}

type acquirePowerAssertionResponse struct {
	HandleID string `json:"handle_id"`
}

type refreshPowerAssertionRequest struct {
	HandleID   string `json:"handle_id"`
	TimeoutSec int    `json:"timeout_sec"`
}

type releasePowerAssertionRequest struct {
	HandleID string `json:"handle_id"`
}

// DevicePowerState is the structured result from /v1/device_power_state (🎯T29).
//
// State values:
//
//	"awake"       — display on, DVT screenshot succeeded, non-trivial pixel content.
//	"display_off" — screenshot succeeded but framebuffer was all-black.
//	"asleep"      — screenshot failed with a pattern indicating device/display off.
//	"unknown"     — prerequisite missing (tunneld down, developer mode off) or
//	                unrecognised error; cannot determine state.
type DevicePowerState struct {
	State  string  `json:"state"`
	Detail *string `json:"detail,omitempty"`
}

type devicePowerStateRequest struct {
	UDID string `json:"udid"`
}

// listDevicesResponse wraps the devices array from the bridge.
type listDevicesResponse struct {
	Devices []DeviceInfo `json:"devices"`
}

// listAppsResponse wraps the apps array from the bridge.
type listAppsResponse struct {
	Apps []AppInfo `json:"apps"`
}
